src/hotspot/share/prims/jvmtiAgent.cpp

```diff
@@ -0,0 +1,620 @@
+// For statically linked agents we cant't rely on os_lib == nullptr because
+// statically linked agents could have a handle of RTLD_DEFAULT which == 0 on some platforms.
+// If this function returns true, then agent->is_static_lib().&& agent->is_loaded().
+static bool load_agent_from_executable(JvmtiAgent* agent, const char* on_load_symbols[], size_t num_symbol_entries) {
+  DEBUG_ONLY(assert_preload(agent);)
+  assert(on_load_symbols != nullptr, "invariant");
+  return os::find_builtin_agent(agent, &on_load_symbols[0], num_symbol_entries);
+}
+
+// Load the library from the absolute path of the agent, if available.
+static void* load_agent_from_absolute_path(JvmtiAgent* agent, bool vm_exit_on_error) {
+  DEBUG_ONLY(assert_preload(agent);)
+  assert(agent->is_absolute_path(), "invariant");
+  assert(!agent->is_instrument_lib(), "invariant");
+  void* const library = os::dll_load(agent->name(), &ebuf[0], sizeof ebuf);
+  if (library == nullptr && vm_exit_on_error) {
+    vm_exit(agent, " in absolute path, with error: ", nullptr);
+  }
+  return library;
+}
+
+// Agents with relative paths are loaded from the standard dll directory.
+static void* load_agent_from_relative_path(JvmtiAgent* agent, bool vm_exit_on_error) {
+  DEBUG_ONLY(assert_preload(agent);)
+  assert(!agent->is_absolute_path(), "invariant");
+  const char* const name = agent->name();
+  void* library = nullptr;
+  // Try to load the agent from the standard dll directory
+  if (os::dll_locate_lib(&buffer[0], sizeof buffer, Arguments::get_dll_dir(), name)) {
+    library = os::dll_load(&buffer[0], &ebuf[0], sizeof ebuf);
+  }
+  if (library == nullptr && os::dll_build_name(&buffer[0], sizeof buffer, name)) {
+    // Try the library path directory.
+    library = os::dll_load(&buffer[0], &ebuf[0], sizeof ebuf);
+    if (library != nullptr) {
+      return library;
+    }
+    if (vm_exit_on_error) {
+      vm_exit(agent, " on the library path, with error: ", missing_module_error_msg);
+    }
+  }
+  return library;
+}
+
+// For absolute and relative paths.
+static void* load_library(JvmtiAgent* agent, const char* on_symbols[], size_t num_symbol_entries, bool vm_exit_on_error) {
+  return agent->is_absolute_path() ? load_agent_from_absolute_path(agent, vm_exit_on_error) :
+                                     load_agent_from_relative_path(agent, vm_exit_on_error);
+}
+
+// Type for the Agent_OnLoad and JVM_OnLoad entry points.
+extern "C" {
+  typedef jint(JNICALL* OnLoadEntry_t)(JavaVM*, char*, void*);
+}
+
+// Find the OnLoad entry point for -agentlib:  -agentpath:   -Xrun agents.
+// num_symbol_entries must be passed-in since only the caller knows the number of symbols in the array.
+static OnLoadEntry_t lookup_On_Load_entry_point(JvmtiAgent* agent, const char* on_load_symbols[], size_t num_symbol_entries) {
+  assert(agent != nullptr, "invariant");
+  if (!agent->is_loaded()) {
+    if (!load_agent_from_executable(agent, on_load_symbols, num_symbol_entries)) {
```

Comment:
It feels like I'm missing something.
We already checked and found at line 322 that `agent->is_loaded() == false`.
Also, we have the comment at line 265:
```
265 // If this function returns true, then agent->is_static_lib().&& agent->is_loaded().
266 static bool load_agent_from_executable(Agent* agent, const char* on_load_symbols[], size_t num_symbol_entries) {
```
As the `agent->is_loaded() == false` then t he condition `agent->is_static_lib() && agent->is_loaded()` has to be `false` and can not be `true`. Then the if-check at line 323 is not needed and can be removed.
Is it right?

Reply:
Good observation, Serguei.

It is because some paths call into lookup_On_load_Entry_point() twice.

It is primarily the attempted conversion of xrun agents, the first invocation comes from JvmtiAgent::convert_xrun_agent(). This will have the agent "loaded". If there is an Agent_OnLoad function, the agent is converted (i.e. xrun removed).

Then when the agent is to invoke the Agent_OnLoad function, there is a second invocation. Here a converted xrun library is already loaded, so I bypass attempting to load it again by checking the is_loaded() property.

Reply:
Thanks.